pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartitionUpsertMetadataManager.java

```diff
@@ -0,0 +1,280 @@
+              String currentSegmentName = currentSegment.getSegmentName();
+              if (segmentName.equals(currentSegmentName)) {
+                if (recordInfo._comparisonValue.compareTo(currentRecordLocation.getComparisonValue()) >= 0) {
+                  validDocIds.add(recordInfo._docId);
+                  return new RecordLocation(segment, recordInfo._docId, recordInfo._comparisonValue);
+                } else {
+                  return currentRecordLocation;
+                }
+              }
+
+              // The current record is in a different segment
+              // Update the record location when getting a newer comparison value, or the value is the same as the
+              // current
+              // value, but the segment has a larger sequence number (the segment is newer than the current segment).
+              if (recordInfo._comparisonValue.compareTo(currentRecordLocation.getComparisonValue()) > 0 || (
+                  recordInfo._comparisonValue == currentRecordLocation.getComparisonValue() && LLCSegmentName
+                      .isLowLevelConsumerSegmentName(segmentName) && LLCSegmentName
+                      .isLowLevelConsumerSegmentName(currentSegmentName)
+                      && LLCSegmentName.getSequenceNumber(segmentName) > LLCSegmentName
+                      .getSequenceNumber(currentSegmentName))) {
+                assert currentSegment.getValidDocIds() != null;
+                currentSegment.getValidDocIds().remove(currentRecordLocation.getDocId());
+                validDocIds.add(recordInfo._docId);
+                return new RecordLocation(segment, recordInfo._docId, recordInfo._comparisonValue);
+              } else {
+                return currentRecordLocation;
+              }
+            } else {
+              // New primary key
+              validDocIds.add(recordInfo._docId);
+              return new RecordLocation(segment, recordInfo._docId, recordInfo._comparisonValue);
+            }
+          });
+    }
+    // Update metrics
+    _serverMetrics.setValueOfPartitionGauge(_tableNameWithType, _partitionId, ServerGauge.UPSERT_PRIMARY_KEYS_COUNT,
+        _primaryKeyToRecordLocationMap.size());
+  }
+
+  /**
+   * Updates the upsert metadata for a new consumed record in the given consuming segment. Returns the merged record if
+   * partial-upsert is enabled.
+   */
+  public GenericRow updateRecord(IndexSegment segment, RecordInfo recordInfo, GenericRow record) {
+    // For partial-upsert, need to ensure all previous records are loaded before inserting new records.
+    if (_partialUpsertHandler != null) {
+      while (!_partialUpsertHandler.isAllSegmentsLoaded()) {
+        LOGGER
+            .info("Sleeping 1 second waiting for all segments loaded for partial-upsert table: {}", _tableNameWithType);
+        try {
+          //noinspection BusyWait
+          Thread.sleep(1000L);
+        } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)